packages/codec/lib/index.ts

```diff
@@ -16,7 +16,147 @@ require("source-map-support/register");
+   * ### Values vs. errors
+   *
+   * It's worth taking a moment here to answer the question: What counts as a value,
+   * and what counts as an error?
+   *
+   * In general, the answer is that anything that can be generated via Solidity
+   * alone (i.e. no assembly), with correctly-encoded inputs, and without making use
+   * of compiler bugs, is a value, not an error.  That means that, for instance, the
+   * following things are values, not errors:
+   *   - A variable of contract type whose address does not actually hold a
+   *     contract of that type;
+   *   - An external function pointer that does not correspond to a valid
+   *     function;
+   *   - A string containing invalid UTF-8;
+   *   - ..., etc.
+   *
+   * By contrast, the following *are* errors:
+   *   - A `bool` which is neither `false` (0) nor `true` (1);
+   *   - An `enum` which is out of range;
+   *   - ..., etc.
+   *
+   * (You may be wondering about the enum case here, because if you go sufficiently
+   * far back, to Solidity 0.4.4 or earlier, it *was* possible to generate
+   * out-of-range enums without resorting to assembly or compiler bugs.  However,
+   * enums are only supported in full mode, which only supports 0.4.9 and later, so
+   * we consider out-of-range enums an error.  There are also additional technical
+   * reasons why supporting out-of-range enums as a value would be difficult.)
+   *
+   * There are three special cases here that are likely worthy of note.
+   *
+   * Firstly, internal function pointers currently can't be meaningfully
+   * decoded via @truffle/decoder.  However, they decode to a bare-bones value,
+   * not an error, as it is (in a sense) our own fault that we can't decode
+   * these, so it doesn't make sense to report an error, which would mean that
+   * something is wrong with the encoded data itself.  This value that it
+   * decodes to will give the program counter values it corresponds to, but
+   * will not include the function name or defining class, as @truffle/decoder
+   * is not presently capable of that.  For now, full decoding of internal
+   * function pointers remains a debugger-only feature.
+   *
+   * (When using the debugger, an invalid internal function pointer will decode to an
+   * error.  However, when using @truffle/decoder, we have no way of discerning whether
+   * the pointer is valid or not, so internal function pointers will always decode to
+   * a value, if an uninformative one.)
+   *
+   * Secondly, when decoding events, it is impossible to decode indexed parameters
+   * of reference type.  Thus, these decode to an error
+   * (`IndexedReferenceTypeError`, which see) rather than to a value.
+   *
+   * Finally, except when decoding events, we do not return an error if the pointers
+   * in an ABI-encoded array or tuple are arranged in a nonstandard way, or if
+   * strings or bytestrings are incorrectly padded, because it is not worth the
+   * trouble to detect these conditions.
+   *
```

Comment:
This is my fault, and possibly belongs in a different PR, but I only just now realized that there's a special case I forgot about when writing this -- overlong arrays / overlarge pointers.  Probably I should take care of that, maybe in a separate PR like I said...

Reply:
Add to #2505?